File: src/main/java/com/tripzin/eleganttex/service/CustomerService.java
package com.tripzin.eleganttex.service;

import com.tripzin.eleganttex.dto.request.CustomerRequest;
import com.tripzin.eleganttex.dto.response.CustomerResponse;
import com.tripzin.eleganttex.entity.Customer;
import com.tripzin.eleganttex.entity.CustomerType;
import com.tripzin.eleganttex.entity.OrderType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Optional;

public interface CustomerService {

    /**
     * Create a new customer
     * @param customerRequest The customer data
     * @return The created customer
     */
    CustomerResponse createCustomer(CustomerRequest customerRequest);
    
    /**
     * Update an existing customer
     * @param id The customer ID
     * @param customerRequest The updated customer data
     * @return The updated customer
     */
    CustomerResponse updateCustomer(Long id, CustomerRequest customerRequest);
    
    /**
     * Get a customer by ID
     * @param id The customer ID
     * @return The customer
     */
    CustomerResponse getCustomerById(Long id);
    
    /**
     * Get all customers with pagination
     * @param pageable Pagination information
     * @return Page of customers
     */
    Page<CustomerResponse> getAllCustomers(Pageable pageable);
    
    /**
     * Search customers by name or phone
     * @param searchTerm The term to search for
     * @param pageable Pagination information
     * @return Page of customers
     */
    Page<CustomerResponse> searchCustomers(String searchTerm, Pageable pageable);
    
    /**
     * Find a customer by phone number
     * @param phone The phone number
     * @return Optional containing the customer if found
     */
    Optional<CustomerResponse> findCustomerByPhone(String phone);
    
    /**
     * Find or create a customer by phone number
     * @param customerRequest The customer data
     * @return The existing or newly created customer
     */
    CustomerResponse findOrCreateCustomer(CustomerRequest customerRequest);
    
    /**
     * Find or create a customer by phone number with order type for customer type assignment
     * @param customerRequest The customer data
     * @param orderType The order type to determine customer type for new customers
     * @return The existing or newly created customer
     */
    CustomerResponse findOrCreateCustomer(CustomerRequest customerRequest, OrderType orderType);
    
    /**
     * Get all customers with pagination and optional customer type filter
     * @param customerType Optional customer type filter
     * @param pageable Pagination information
     * @return Page of customers
     */
    Page<CustomerResponse> getAllCustomers(CustomerType customerType, Pageable pageable);
    
    /**
     * Search customers by name or phone with optional customer type filter
     * @param searchTerm The term to search for
     * @param customerType Optional customer type filter
     * @param pageable Pagination information
     * @return Page of customers
     */
    Page<CustomerResponse> searchCustomers(String searchTerm, CustomerType customerType, Pageable pageable);
    
    /**
     * Delete a customer
     * @param id The customer ID
     */
    void deleteCustomer(Long id);
    
    /**
     * Get a customer entity by ID
     * @param id The customer ID
     * @return The customer entity
     */
    Customer getCustomerEntityById(Long id);
}

File: frontend/src/services/customer.service.ts
import api from './api';
import { Customer, CustomerRequest, CustomerType } from '../types/customer';

const BASE_URL = '/customers';

/**
 * Get all customers with pagination and optional customer type filter
 * @param page Page number (0-based)
 * @param size Page size
 * @param customerType Optional customer type filter
 * @returns Paginated list of customers
 */
export const getCustomers = async (page = 0, size = 10, customerType?: CustomerType): Promise<{ content: Customer[], totalElements: number }> => {
  try {
    const params: any = { page, size };
    if (customerType) {
      params.customerType = customerType;
    }
    
    const response = await api.get(BASE_URL, { params });
    return response.data;
  } catch (error) {
    console.error('Error fetching customers:', error);
    throw error;
  }
};

/**
 * Get a customer by ID
 * @param id Customer ID
 * @returns Customer details
 */
export const getCustomerById = async (id: number): Promise<Customer> => {
  try {
    const response = await api.get(`${BASE_URL}/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching customer with ID ${id}:`, error);
    throw error;
  }
};

/**
 * Search customers by name or phone with optional customer type filter
 * @param term Search term
 * @param page Page number (0-based)
 * @param size Page size
 * @param customerType Optional customer type filter
 * @returns Paginated list of customers matching the search term
 */
export const searchCustomers = async (term: string, page = 0, size = 10, customerType?: CustomerType): Promise<{ content: Customer[], totalElements: number }> => {
  try {
    const params: any = { term, page, size };
    if (customerType) {
      params.customerType = customerType;
    }
    
    const response = await api.get(`${BASE_URL}/search`, { params });
    return response.data;
  } catch (error) {
    console.error('Error searching customers:', error);
    throw error;
  }
};

/**
 * Find a customer by phone number
 * @param phone Phone number
 * @returns Customer details if found, null otherwise
 */
export const findCustomerByPhone = async (phone: string): Promise<Customer | null> => {
  try {
    const response = await api.get(`${BASE_URL}/phone/${phone}`);
    return response.data;
  } catch (error: any) {
    if (error.status === 404) {
      return null; // Customer not found
    }
    console.error(`Error finding customer with phone ${phone}:`, error);
    throw error;
  }
};

/**
 * Create a new customer
 * @param customerData Customer data
 * @returns Created customer
 */
export const createCustomer = async (customerData: CustomerRequest): Promise<Customer> => {
  try {
    const response = await api.post(BASE_URL, customerData);
    return response.data;
  } catch (error) {
    console.error('Error creating customer:', error);
    throw error;
  }
};

/**
 * Update an existing customer
 * @param id Customer ID
 * @param customerData Updated customer data
 * @returns Updated customer
 */
export const updateCustomer = async (id: number, customerData: CustomerRequest): Promise<Customer> => {
  try {
    const response = await api.put(`${BASE_URL}/${id}`, customerData);
    return response.data;
  } catch (error) {
    console.error(`Error updating customer with ID ${id}:`, error);
    throw error;
  }
};

/**
 * Find an existing customer by phone or create a new one
 * @param customerData Customer data
 * @returns Existing or newly created customer
 */
export const findOrCreateCustomer = async (customerData: CustomerRequest): Promise<Customer> => {
  try {
    const response = await api.post(`${BASE_URL}/find-or-create`, customerData);
    return response.data;
  } catch (error) {
    console.error('Error finding or creating customer:', error);
    throw error;
  }
};

/**
 * Delete a customer
 * @param id Customer ID
 */
export const deleteCustomer = async (id: number) => {
  try {
    await api.delete(`${BASE_URL}/${id}`);
  } catch (error) {
    console.error(`Error deleting customer with ID ${id}:`, error);
    throw error;
  }
};

const customerService = {
  getCustomers,
  getCustomerById,
  searchCustomers,
  findCustomerByPhone,
  createCustomer,
  updateCustomer,
  findOrCreateCustomer,
  deleteCustomer
};

export default customerService;
